src/timer/pwm.rs

```diff
@@ -108,15 +108,50 @@ macro_rules! pwm {
                     let psc = (ratio - 1) / 0xffff;
                     let arr = ratio / (psc + 1) - 1;
 
+                    self.set_prescaler_register(psc as u16);
+                    self.set_period_register(arr as u16);
+
                     unsafe {
-                        self.tim.psc.write(|w| w.psc().bits(psc as u16));
-                        self.tim.arr.write(|w| w.$arr().bits(arr as u16));
+                        self.tim.cr1.write(|w| w.cen().set_bit());
+                    }
+                }
+
+                /// Set only period register
+                pub fn set_period_register(&mut self, period: u16) {
+                    unsafe {
+                        self.tim.arr.write(|w| w.$arr().bits(period as u16));
                         $(
-                            self.tim.arr.modify(|_, w| w.$arr_h().bits((arr >> 16) as u16));
+                            self.tim.arr.modify(|_, w| w.$arr_h().bits((period >> 16) as u16));
```

Comment:
```rust
pwm! {
    TIM1: (tim1, arr),
    TIM3: (tim3, arr_l, arr_h), // should this really have a `arr_h`?
    TIM14: (tim14, arr),
    TIM16: (tim16, arr),
    TIM17: (tim17, arr),
}
```

I must say that I am a bit confused about the 16/32bit for TIM3. From what I understand from the [reference manual for g0x0](https://www.st.com/resource/en/reference_manual/rm0454-stm32g0x0-advanced-armbased-32bit-mcus-stmicroelectronics.pdf#page=434) TIM3 is just 16-bit. [reference manual for g0x1](https://www.st.com/resource/en/reference_manual/dm00371828-stm32g0x1-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf#page=625) says pretty much the same thing except it adds a TIM2 which *is* 32-bit.

Looking at the pac, TIM3 does have an `arr_h`(inherited from tim2?), same thing when looking in the reference manual...

Also is it really safe to write to `arr` in two steps without the pwm signal going crazy in between?

Reply:
Please correct me if I am wrong but from what I can gather, TIM3 is 16bits but does share the same memory config structure as TIM2(which is 32bit)